design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,138 @@
+
+Add a new Custom Resource(CR), `DataImage`, to the cluster which will make it possible,
+using ownerReferences (example : [HostFirmwareSettings](https://github.com/metal3-io/baremetal-operator/blob/9ce684e6462a6ab55ff650cb6c11f9ba1ffb395d/docs/api.md?plain=1#L664)) and same `Name` and `Namespace` as the `BareMetalHost`, to attach a generic non-bootable ISO (a CD "ISO 9660" image) image via Ironic, after the host has been provisioned.
+
+## Motivation
+
+In certain scenarios it is desirable to attach a non-bootable iso image
+to a system after provisioning has finished, for example :
+
+A configuration ISO can be used to provide first-boot configuration to a host
+deployed from a standard image in advance. An operator may have a pool of
+powered off hosts with operating system already provisioned. Then, adding
+a host to the cluster boils down to powering it on with a configuration ISO provided.
+
+### Goals
+
+Expose the Ironic API to attach non-bootable iso images via Metal3.
+
+### Non-Goals
+
+Supporting non-ISO images i.e. media types other than CD/DVD, supporting
+drivers other than Redfish and its derived drivers.
+
+## Proposal
+
+### BMO Proposal
+
+Add a new Custom Resource(CR), `DataImage` (with same `name` and `namespace` as the BMH), which contains the spec containing details of the
+non-bootable iso image, example :
+
+#### Design spec
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: DataImage
+metadata:
+  name: metal-worker-0
+  namespace: metal-cluster
+spec:
+  dataImage:
+    url: http://1.2.3.4/image.iso
+```
+
+We will use ownerReferences in the `DataImage` CR to link it to the corresponding `BareMetalHost` object.
+
+Since we are using a CR, a user can create RBAC policies that doesn't allow unauthorized users
+to attach a non-bootable iso image to hosts.
+
+We may consider adding support for HTTP credentials and TLS settings in the future
+but it's outside the scope of this proposal.
+
+**Note** : As part of this proposal, the dataImage will be attached after the node
+has been Provisioned or ExternallyProvisioned.
+
+### Ironic Support
+
+The [proposal](https://bugs.launchpad.net/ironic/+bug/2033288) for exposing this functionaility in Ironic has been
+accepted by the upstream Ironic community.
+
+The [Ironic implementation](https://review.opendev.org/c/openstack/ironic/+/894918) based on the above proposal has been completed.
+
+Other than that, we just need the implementation for driver specific support,
+which in this proposal is Redfish.
+
+## Design Details
+
+Create a new Custom Resource(CR), `DataImage`, with spec field `Url` and same `Name`
+and `Namespace` as the `BareMetalHost` where we are attaching the image.
+
+The ISO will be attached when the `DataImage` CR object is created and the
+corresponding ownerReference is added for the `BareMetalHost` object,  which will trigger a
+reconciliation. In case a reboot is requested at the same time using
+`RebootAnnotation`, the ISO will be attached first.
```

Comment:
There is no "at the same time" any more: we have two different resources. Now I'd expect the DataImage to be created first, but only handled by the BMH controller on reboot/power on.

Reply:
Right, I have updated the proposal.